autogen/graph_utils.py

```diff
@@ -0,0 +1,128 @@
+from typing import Dict, List, Optional, Union
+import logging
+
+from autogen.agentchat.groupchat import Agent
+
+
+def check_graph_validity(
+    allowed_speaker_order_dict: dict,
+    agents: List[Agent],
+    allow_repeat_speaker: Optional[Union[bool, List[Agent]]] = True,
+):
+    """
+    allowed_speaker_order_dict: A dictionary of keys and list as values. The keys are the names of the agents, and the values are the names of the agents that the key agent can transition to.
+    agents: A list of Agents
+    allow_repeat_speaker: A boolean indicating whether the same agent can speak twice in a row.
```

Comment:
Hey @sonichi, before I work on the request, can we expand on this point?

1. `allow_repeat_speaker` is currently used as a check against `speaker_order_dict`. To be specific, if `allow_repeat_speaker` is True, then, self-loops are allowed in `speaker_order_dict` for any agent. If `allow_repeat_speaker` is a list of Agents, then, self-loops are allowed only for the named agents in `speaker_order_dict`.

2. I would recommend for `allow_repeat_speaker` not to be ignored because I think that taking one variable as ground-truth (`speaker_order_dict`) and ignoring another parameter (`allow_repeat_speaker`) causes silent failures/ surprises to the user.

3. Great pickup on Lines 86-90, I will refactor Line 86-90 to take in `allow_repeat_speaker` if we agree on [2].

Curious to hear your thoughts. :)

Reply:
I don't mean make it a silent failure. We can explicitly raise an error when both of them are provided. For users they don't really need to provide both. In fact, I view "allow_repeat_speaker" as a shortcut to specify a special graph. `speaker_order_dict` is a more general way to specify a graph. They shouldn't be used together. For devs, that simplifies things too. 
Currently allow_repeat_speaker is set to True by default. We can use a `Sentinel` like #831 to handle this problem. If `speaker_order_dict` is not given, then its default is True; otherwise, its default is None.

Reply:
Implemented as per above discussion.